src/main/kotlin/org/rust/ide/annotator/RsFormatMacroAnnotator.kt

```diff
@@ -0,0 +1,127 @@
+
+    override fun annotateInternal(element: PsiElement, holder: AnnotationHolder) {
+        val formatMacro = element as? RsMacroCall ?: return
+
+        val formatStr = formatMacro
+            .formatMacroArgument
+            ?.formatMacroArgList
+            ?.getOrNull(macroToFormatPos(formatMacro.macroName))
+            ?.descendantOfTypeStrict<RsLitExpr>()
+            ?: return
+
+        val literalKind = (formatStr.kind as? RsLiteralKind.String) ?: return
+        val rawTextRange = literalKind.offsets.value ?: return
+        val (unescapedText, sourceMap, _) = parseRustStringCharacters(rawTextRange.substring(literalKind.node.text))
+
+        val arguments = formatParser.findAll(unescapedText.toString())
+        var currentOffset = 0
+
+        fun IntRange.toSourceRange(additionalOffset: Int = currentOffset): TextRange =
+            TextRange(sourceMap[first + additionalOffset], sourceMap[last + additionalOffset] + 1)
+                .shiftRight(formatStr.startOffset + rawTextRange.startOffset)
+
+        fun AnnotationHolder.highlightArgument(range: IntRange?, color: RsColor = RsColor.IDENTIFIER) {
+            if (range != null && !range.isEmpty())
+                createAnnotation(
+                    ARG_HIGHLIGHTING,
+                    range.toSourceRange(),
+                    null)
+                    .textAttributes = color.textAttributesKey
+        }
+
+        val key = RsColor.VALID_STRING_ESCAPE
+        val highlightSeverity = if (isUnitTestMode) key.testSeverity else HighlightSeverity.INFORMATION
+        for (arg in arguments) {
+            currentOffset = 0
+            val textRange = arg.range.toSourceRange()
+            holder.createAnnotation(highlightSeverity, textRange, null)
+                .textAttributes = key.textAttributesKey
+
+            if (arg.groups[3] != null) {
+                holder.createErrorAnnotation(
+                    arg.range.toSourceRange(),
+                    "Invalid format string: unmatched '}'")
+            }
+            if (arg.groups[1] != null) {
+
+                val inside = arg.groups[2]!!
+                currentOffset = inside.range.first
+                val matchResult = formatArgParser.find(inside.value)!!
+
+                if (!arg.groups[1]!!.value.endsWith("}")) {
+                    val possibleEnd = matchResult.value.length - 1
+                    holder.createErrorAnnotation(
+                        (possibleEnd..possibleEnd).toSourceRange(),
+                        "Invalid format string: } expected.\nIf you intended to print `{` symbol, you can escape it using `{{`")
+                    continue
+                }
+
+                val validParsedEnd = matchResult.range.last + 1
+                if (validParsedEnd != inside.value.length) {
+                    holder.createErrorAnnotation(
+                        (validParsedEnd until inside.value.length).toSourceRange(),
+                        "Invalid format string")
```

Comment:
And here `)`

Reply:
Sorry, for the amount of code style issues. Rustfmt is very aggressive in formatting(sometimes even too aggressive), Kotlin formatter is usually also enough, so usually don't even think about it.